x/fundraising/keeper/auction.go

```diff
@@ -221,3 +222,76 @@ func (k Keeper) CancelAuction(ctx sdk.Context, msg *types.MsgCancelAuction) (typ
 
 	return auction, nil
 }
+
+// AddAllowedBidders is a function for an external module and it simply adds new bidder(s) to AllowedBidder list.
+// Note that it doesn't do auctioneer verification because the module is generalized for broader use cases.
+// It is designed to delegate to an external module to add necessary verification and logics depending on their use case.
+func (k Keeper) AddAllowedBidders(ctx sdk.Context, auctionId uint64, bidders []types.AllowedBidder) error {
+	auction, found := k.GetAuction(ctx, auctionId)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "auction %d is not found", auctionId)
+	}
+
+	if len(bidders) == 0 {
+		return types.ErrEmptyAllowedBidders
+	}
+
+	for _, bidder := range bidders {
+		if bidder.MaxBidAmount.IsNil() {
+			return types.ErrInvalidMaxBidAmount
+		}
+
+		if !bidder.MaxBidAmount.IsPositive() {
+			return types.ErrInvalidMaxBidAmount
+		}
+	}
+
+	// Append new bidders from the existing ones
+	allowedBidders := auction.GetAllowedBidders()
+	allowedBidders = append(allowedBidders, bidders...)
+
+	if err := auction.SetAllowedBidders(allowedBidders); err != nil {
+		return err
+	}
+	k.SetAuction(ctx, auction)
+
+	return nil
+}
+
+// UpdateAllowedBidder is a function for an external module and it simply updates the bidder's maximum bid amount.
+// Note that it doesn't do auctioneer verification because the module is generalized for broader use cases.
+// It is designed to delegate to an external module to add necessary verification and logics depending on their use case.
+func (k Keeper) UpdateAllowedBidder(ctx sdk.Context, auctionId uint64, bidder sdk.AccAddress, maxBidAmount sdk.Int) error {
+	auction, found := k.GetAuction(ctx, auctionId)
+	if !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "auction %d is not found", auctionId)
+	}
+
+	allowedBiddersMap := make(map[string]sdk.Int)
+	for _, bidder := range auction.GetAllowedBidders() {
+		allowedBiddersMap[bidder.GetBidder()] = bidder.MaxBidAmount
+	}
+
+	if _, found := allowedBiddersMap[bidder.String()]; !found {
+		return sdkerrors.Wrapf(sdkerrors.ErrNotFound, "bidder %s is not found", bidder.String())
+	}
+
+	if maxBidAmount.IsNil() {
+		return types.ErrInvalidMaxBidAmount
+	}
+
+	if !maxBidAmount.IsPositive() {
+		return types.ErrInvalidMaxBidAmount
+	}
```

Comment:
I would have add these conditions first to avoid iterating bidders in case input is invalid

Reply:
I will move them up